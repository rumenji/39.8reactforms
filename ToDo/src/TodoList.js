import React, { useState } from "react";
import Todo from "./Todo";
import NewTodoForm from "./NewTodoForm";
import EditTodoForm from "./EditTodoForm";
import { v4 as uuid } from "uuid";

function TodoList(){
    const [todos, setTodos] = useState([]);

    const addTodo = todo => {
        const newTodo = {...todo, id: uuid(), completed: false};
        setTodos(todos => [...todos, newTodo])
    }

    const removeTodo = id => {
        const todosCopy = [...todos];
        setTodos(todos => todosCopy.filter(todo => todo.id !== id))
    }
    
    const [editTodoForm, setEditTodoForm] = useState();

    const showEditForm = (id, task) => {
        setEditTodoForm(<EditTodoForm id={id} task={task} editTodo={editTodo} />)
    }

    const editTodo = (id, task) => {
        const todosCopy = [...todos];
        todosCopy.map(todo => {
            if(todo.id === id){
                todo.task = task.task;

            }
            return todo;
        })
        setEditTodoForm();
        setTodos(todosCopy);
    }

    const completeTodo = (id) => {
        const todosCopy = [...todos];
        todosCopy.map(todo => {
            if(todo.id === id){
                todo.completed = !todo.completed;

            }
            console.log(todo)
            return todo;
        })
        setTodos(todosCopy);
    };

    return (
        <div className="TodosList">
            <NewTodoForm addTodo={addTodo} />
            {editTodoForm}
            {todos.map((todo) => (<Todo 
                                   key={todo.id} 
                                   task={todo.task} 
                                   completed = {todo.completed}
                                   complete = {()=> completeTodo(todo.id)}
                                   remove={() => removeTodo(todo.id)} 
                                   edit={() => showEditForm(todo.id, todo.task)}
                                   testId={todo.id}/>))}
        </div>
    )
}

export default TodoList;