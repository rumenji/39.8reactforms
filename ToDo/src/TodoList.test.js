import React from "react";
import { render, fireEvent } from '@testing-library/react';
import TodoList from "./TodoList";

it("renders without crashing", function(){
    render(<TodoList />)
})

it("matches snapshot", function(){
    const {asFragment} = render(<TodoList />);
    expect(asFragment()).toMatchSnapshot();
})

it("can add new todos", function(){
    const {getByLabelText, queryByText} = render(<TodoList />);

    const todoInput = getByLabelText("Enter to-do:");
    const submitBtn = queryByText("Add to-do")

    fireEvent.change(todoInput, {target: {value: "Pickup"}});

    fireEvent.click(submitBtn);

    expect(queryByText("Pickup")).toBeInTheDocument();
})

it("can remove todos", function(){
    const {getByLabelText, queryByText} = render(<TodoList />);

    const todoInput = getByLabelText("Enter to-do:");
    const submitBtn = queryByText("Add to-do")

    fireEvent.change(todoInput, {target: {value: "Pickup"}});

    fireEvent.click(submitBtn);

    expect(queryByText("Pickup")).toBeInTheDocument();

    const removeBtn = queryByText("X");
    fireEvent.click(removeBtn);

    expect(queryByText("Pickup")).not.toBeInTheDocument();

})