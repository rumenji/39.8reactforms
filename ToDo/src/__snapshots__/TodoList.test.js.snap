// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="TodosList"
  >
    <form>
      <label
        for="task"
      >
        Enter to-do:
      </label>
      <input
        id="task"
        name="task"
        placeholder="Enter to-do task"
        type="text"
        value=""
      />
      <button>
        Add to-do
      </button>
    </form>
  </div>
</DocumentFragment>
`;
